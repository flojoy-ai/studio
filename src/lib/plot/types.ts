export interface Drawable {
  draw: () => void;
}
