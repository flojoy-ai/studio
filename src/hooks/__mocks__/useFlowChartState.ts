export const useFlowChartState = jest.fn(() => ({
  rfInstance: {
    nodes: [],
  },
  nodes: [],
  edges: [],
  updateCtrlInputDataForNode: jest.fn(),
  removeCtrlInputDataForNode: jest.fn(),
  ctrlsManifest: [
    {
      type: "input",
      name: "Slider",
      id: "INPUT_PLACEHOLDER",
      hidden: false,
      minHeight: 1,
      minWidth: 2,
    },
  ],
  setCtrlsManifest: jest.fn(),
  isEditMode: false,
  gridLayout: [],
  setGridLayout: jest.fn(),
  setRfInstance: jest.fn(),
  loadFlowExportObject: jest.fn(),
  openFileSelector: jest.fn(),
  saveFile: jest.fn(),
  setIsEditMode: jest.fn(),
  uiTheme: "dark",
  setUiTheme: jest.fn(),
  showLogs: false,
  setShowLogs: jest.fn(),
  runningNode: "",
  setRunningNode: jest.fn(),
  failedNode: "",
  setFailedNode: jest.fn(),
  setEdges: jest.fn(),
  setNodes: jest.fn(),
}));
