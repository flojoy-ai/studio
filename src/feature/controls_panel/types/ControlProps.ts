import { ResultsType } from "@src/feature/results_panel/types/ResultsType";
import { CtlManifestType } from "@src/hooks/useFlowChartState";
import { Dispatch, SetStateAction } from "react";
import { CtrlOptionValue } from "./ControlOptions";

export interface ControlProps {
  isEditMode: boolean;
  results: ResultsType;
  updateCtrlValue: (val: string, ctrl: CtlManifestType) => void;
  attachParamsToCtrl: (param: CtrlOptionValue, ctrl: CtlManifestType) => void;
  removeCtrl: (
    e: React.MouseEvent<HTMLButtonElement, MouseEvent>,
    ctrl: CtlManifestType
  ) => void;
  setCurrentInput: Dispatch<
    SetStateAction<
      | (CtlManifestType & {
          index: number;
        })
      | undefined
    >
  >;
  setOpenEditModal: Dispatch<SetStateAction<boolean>>;
}
