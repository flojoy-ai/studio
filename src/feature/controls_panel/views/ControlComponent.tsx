import { Dispatch, memo, SetStateAction, useCallback } from "react";
import Select, { ThemeConfig } from "react-select";
import Slider from "rc-slider";
import "rc-slider/assets/index.css";
import customDropdownStyles from "../style/CustomDropdownStyles";

import { ControlNames } from "../manifest/CONTROLS_MANIFEST";
import ControlComponentState from "./ControlComponentState";
import useControlComponentEffects from "@hooks/useControlComponentEffects";
import {
  CtlManifestType,
  CtrlManifestParam,
  PlotManifestParam,
} from "@src/hooks/useFlowChartState";
import { ResultsType } from "@src/feature/results_panel/types/ResultsType";
import { CtrlOptionValue } from "../types/ControlOptions";
import PlotControl from "./PlotControl";
import SevenSegmentComponent from "./SevenSegmentComponent";
import KnobCtrl from "./KnobCtrl";
import NodeReference from "./NodeReference";

export type ControlComponentProps = {
  ctrlObj: CtlManifestType;
  theme: "light" | "dark";
  results: ResultsType;
  updateCtrlValue: (value: string, ctrl: CtlManifestType) => void;
  attachParamsToCtrl: (
    val: CtrlOptionValue | PlotManifestParam,
    ctrlObj: CtlManifestType
  ) => void;
  removeCtrl: (
    e: React.MouseEvent<HTMLButtonElement, MouseEvent>,
    ctrl: CtlManifestType
  ) => void;
  setCurrentInput: Dispatch<
    SetStateAction<CtlManifestType & { index: number }>
  >;
  setOpenEditModal: Dispatch<SetStateAction<boolean>>;
};

const ControlComponent = ({
  ctrlObj,
  theme,
  results,
  updateCtrlValue,
  attachParamsToCtrl,
  removeCtrl,
  setCurrentInput,
  setOpenEditModal,
}: ControlComponentProps) => {
  const {
    ctrlsManifest,
    setGridLayout,
    isEditMode,
    selectOptions,
    setSelectOptions,
    flowChartObject,
    textInput,
    setTextInput,
    numberInput,
    setNumberInput,
    sliderInput,
    setSliderInput,
    currentInputValue,
    setCurrentInputValue,
    nd,
    setNd,
    setPlotData,
    selectedPlotOption,
    plotData,
    selectedOption,
    setSelectedOption,
    inputOptions,
    outputOptions,
    setOutputOptions,
    setSelectedPlotOption,
    ctrls,
    defaultValue,
    paramOptions,
    styledLayout,
    openFileSelector,
  } = ControlComponentState({
    updateCtrlValue,
    theme,
    ctrlObj,
  });

  const handleCtrlValueChange = (
    setValue: Dispatch<SetStateAction<string>>,
    value: string
  ) => {
    setValue(value);
    if (!(ctrlObj?.param as CtrlManifestParam)?.nodeId) {
      return;
    }
    updateCtrlValue(value, ctrlObj);

    if ((ctrlObj.param as CtrlManifestParam).functionName === "CONSTANT") {
      attachParamsToCtrl(
        {
          ...(selectedOption?.value as CtrlOptionValue),
          id: `CONSTANT_${value.toString().split(" ").join("")}_constant`,
        },
        { ...ctrlObj, val: value }
      );
    }
  };

  const makeLayoutStatic = () => {
    if (isEditMode) {
      setGridLayout((prev) => {
        prev[prev.findIndex((layout) => layout.i === ctrlObj.id)].static = true;
      });
    }
  };

  useControlComponentEffects({
    nd,
    selectedPlotOption,
    setNd,
    setPlotData,
    ctrlObj,
    ctrls,
    ctrlsManifest,
    currentInputValue,
    defaultValue,
    flowChartObject,
    isEditMode,
    numberInput,
    paramOptions,
    plotData,
    results,
    selectedOption,
    selectOptions,
    inputOptions,
    outputOptions,
    sliderInput,
    styledLayout,
    textInput,
    setCurrentInputValue,
    setGridLayout,
    setNumberInput,
    setSelectedOption,
    setSelectedPlotOption,
    setSelectOptions,
    setOutputOptions,
    setSliderInput,
    setTextInput,
    openFileSelector,
  });
  return (
    <div
      style={{
        display: "flex",
        flexDirection: "column",
        justifyContent: "center",
        alignItems: "center",
        flex: "1",
        padding: "16px",
      }}
      data-cy={ctrlObj.id}
    >
      {isEditMode && (
        <div
          className="ctrl-header"
          data-cy="ctrls-select"
          style={{ width: "100%" }}
        >
          <Select
            className="select-node"
            isSearchable={true}
            onChange={(val) => {
              if (val)
                attachParamsToCtrl(val.value as CtrlOptionValue, ctrlObj);
            }}
            theme={theme as unknown as ThemeConfig}
            options={selectOptions}
            styles={customDropdownStyles}
            value={selectedOption}
            inputId={`select-input-${ctrlObj.id}`}
          />
          <button
            className="ctrl-edit-btn"
            onClick={() => {
              setCurrentInput({
                ...ctrlObj,
                index: ctrlsManifest.findIndex(
                  (manifest) => manifest.id === ctrlObj.id
                ),
              });
              setOpenEditModal(true);
            }}
          >
            &#9998;
          </button>

          <button
            onClick={(e) => removeCtrl(e, ctrlObj)}
            id={ctrlObj.id}
            className="ctrl-edit-btn"
          >
            x
          </button>
        </div>
      )}
      {!isEditMode && (
        <p className="ctrl-param">Node: {selectedOption?.label}</p>
      )}
      {ctrlObj.name === ControlNames.Plot && (
        <PlotControl
          nd={nd}
          results={results}
          setNd={setNd}
          setPlotData={setPlotData}
          ctrlObj={ctrlObj}
          isEditMode={isEditMode}
          plotData={plotData}
          selectedOption={selectedOption}
          selectedPlotOption={selectedPlotOption}
          theme={theme}
          setSelectedPlotOption={setSelectedPlotOption}
        />
      )}
      {ctrlObj.name === ControlNames.LocalFileLoader && (
        <div className="ctrl-input-body-file" data-cy="numeric-input">
          <input
            type={"text"}
            placeholder={"Please enter the full file path"}
            className="ctrl-numeric-input border-color"
            onChange={(e) => {
              handleCtrlValueChange(setTextInput, e.target.value);
            }}
            value={currentInputValue || textInput || ""}
            style={{ ...(theme === "dark" && { color: "#fff" }) }}
          />
          {/* <button
            className="cmd-btn-dark "
            style={{
              ...{color: theme === "dark" ? "#fff" : "#000"}
            }}
            onClick={openFileSelector}
            >
            Select File
          </button> */}
        </div>
      )}
      {ctrlObj.name === ControlNames.SevenSegmentDisplay && (
        <SevenSegmentComponent ctrlObj={ctrlObj} plotData={plotData} nd={nd} />
      )}

      {ctrlObj.name === ControlNames.TextInput && (
        <div className="ctrl-input-body" data-cy="numeric-input">
          <input
            type="text"
            placeholder="Write your text.."
            className="ctrl-numeric-input border-color"
            onChange={(e) => {
              handleCtrlValueChange(setTextInput, e.target.value);
            }}
            value={currentInputValue || textInput || ""}
            style={{ ...(theme === "dark" && { color: "#fff" }) }}
          />
        </div>
      )}

      {ctrlObj.name === ControlNames.NumericInput && (
        <div className="ctrl-input-body" data-cy="numeric-input">
          <input
            type="number"
            placeholder="Enter a number"
            className="ctrl-numeric-input border-color"
            onChange={(e) => {
              handleCtrlValueChange(setNumberInput, e.target.value);
            }}
            value={currentInputValue || numberInput || 0}
            style={{ ...(theme === "dark" && { color: "#fff" }) }}
          />
        </div>
      )}

      {ctrlObj.name === ControlNames.StaticNumericInput && (
        <div className="ctrl-input-body">
          <input
            type="number"
            placeholder="Enter a number"
            className="ctrl-numeric-input"
            onChange={(e) => {
              handleCtrlValueChange(setNumberInput, e.target.value);
            }}
            disabled
            value={currentInputValue || numberInput || 0}
          />
        </div>
      )}

      {ctrlObj.name === ControlNames.Knob && (
        <div className="ctrl-input-body">
          <KnobCtrl
            selectedOption={selectedOption}
            ctrlObj={ctrlObj}
            isEditMode={isEditMode}
            makeLayoutStatic={makeLayoutStatic}
            setGridLayout={setGridLayout}
            updateCtrlValue={updateCtrlValue}
            currentInputValue={+currentInputValue}
          />
        </div>
      )}

      {ctrlObj.name === ControlNames.Slider && (
        <div className="ctrl-input-body">
          <div
            onMouseEnter={makeLayoutStatic}
            onMouseLeave={() => {
              if (isEditMode) {
                setGridLayout((prev) => {
                  prev[
                    prev.findIndex((layout) => layout.i === ctrlObj.id)
                  ].static = false;
                });
              }
            }}
            style={{ width: "100%" }}
          >
            <Slider
              className="custom-slider"
              onChange={(val) => {
                handleCtrlValueChange(setSliderInput, val.toString());
              }}
              value={+currentInputValue || +sliderInput || 0}
            />
            <label>{currentInputValue ? currentInputValue : sliderInput}</label>
          </div>
        </div>
      )}
      {ctrlObj.name === ControlNames.NodeReference && (
        <NodeReference
          ctrlObj={ctrlObj}
          updateCtrlValue={updateCtrlValue}
          theme={theme}
        />
      )}
      {ctrlObj.name === ControlNames.Dropdown && (
        <div className="ctrl-input-body">
          <Select
            className="select-node"
            isSearchable={true}
            onChange={(val) => {
              updateCtrlValue(
                (val as { label: string; value: string }).value,
                ctrlObj
              );
            }}
            theme={theme as unknown as ThemeConfig}
            options={paramOptions}
            styles={customDropdownStyles}
            value={
              paramOptions.find(
                (opt) => opt.value === currentInputValue.toString()
              ) || { label: "", value: "" }
            }
          />
        </div>
      )}

      {ctrlObj.name === ControlNames.CheckboxButtonGroup && (
        <div className="ctrl-input-body">
          {paramOptions.map((option) => {
            return (
              <div key={option.value}>
                <input
                  type="checkbox"
                  id={`${ctrlObj.id}_${option.value}`}
                  name={`${ctrlObj.id}_${option.value}`}
                  value={option.value}
                  checked={currentInputValue.toString() === option.value}
                  onChange={(e) => {
                    updateCtrlValue(option.value, ctrlObj);
                  }}
                />
                <label htmlFor={`${ctrlObj.id}_${option.value}`}>
                  {" "}
                  {option.label}{" "}
                </label>
              </div>
            );
          })}
        </div>
      )}

      {ctrlObj.name === ControlNames.RadioButtonGroup && (
        <div className="ctrl-input-body">
          {paramOptions.map((option) => {
            return (
              <div style={{ width: "max-content" }} key={option.value}>
                <input
                  type="radio"
                  id={`${ctrlObj.id}_${option.value}`}
                  name={`${ctrlObj.id}_${option.value}`}
                  value={option.value}
                  checked={currentInputValue.toString() === option.value}
                  onChange={(e) => {
                    updateCtrlValue(option.value, ctrlObj);
                  }}
                />
                <label htmlFor={`${ctrlObj.id}_${option.value}`}>
                  {" "}
                  {option.label}{" "}
                </label>
              </div>
            );
          })}
        </div>
      )}

      <details className="ctrl-meta">
        {`Name: ${ctrlObj.name}`}
        <br></br>
        {`ID: ${ctrlObj.id}`}
      </details>
    </div>
  );
};
export default memo(ControlComponent);
