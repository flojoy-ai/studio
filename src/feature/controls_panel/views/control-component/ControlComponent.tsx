import Slider from "rc-slider";
import "rc-slider/assets/index.css";
import { Dispatch, memo, SetStateAction } from "react";
import Select, { ThemeConfig } from "react-select";
import customDropdownStyles from "../../style/CustomDropdownStyles";

import useControlComponentEffects from "@hooks/useControlComponentEffects";
import {
  Box,
  clsx,
  createStyles,
  Text,
  useMantineColorScheme,
} from "@mantine/core";
import { ResultsType } from "@src/feature/results_panel/types/ResultsType";
import {
  CtlManifestType,
  CtrlManifestParam,
} from "@src/hooks/useFlowChartState";
import { ControlNames } from "../../manifest/CONTROLS_MANIFEST";
import { CtrlOptionValue } from "../../types/ControlOptions";
import KnobCtrl from "../KnobCtrl";
import NodeReference from "../NodeReference";
import PlotControl from "../PlotControl";
import SevenSegmentComponent from "../SevenSegmentComponent";
import ControlComponentState from "./ControlComponentState";

export const useControlStyles = createStyles((theme) => {
  return {
    header: {
      width: "100%",
      display: "flex",
      justifyContent: "flex-end",
      alignItems: "center",
    },
    btn: {
      cursor: "pointer",
      background: "transparent",
      border: "none",
      margin: 0,
      padding: 5,
      borderRadius: theme.radius.sm,
      "&:hover": {
        backgroundColor:
          theme.colorScheme === "dark"
            ? theme.colors.dark[5]
            : theme.colors.gray[3],
      },
    },
    closeBtn: {
      "&:hover": {
        color: theme.colors.red[7],
      },
    },
    inputBody: {
      padding: 20,
      display: "flex",
      justifyContent: "center",
      alignItems: "center",
      height: "90%",
      flexWrap: "wrap",
      width: "100%",
    },
    inputBodyFile: {
      display: "grid",
      justifyContent: "center",
      alignItems: "center",
      height: "90%",
      flexWrap: "wrap",
      width: "100%",
    },
    param: {
      margin: 0,
      fontSize: 15,
    },
    meta: {
      fontSize: 10,
      textAlign: "left",
      position: "absolute",
      bottom: 10,
      left: 15,
    },
    ctrlNumericInput: {
      padding: 10,
      width: "90%",
      margin: "0 10px",
      backgroundColor: "transparent",
      outline: "none",
      border: `1px solid ${theme.colors.gray[3]}`,
      borderRadius: 10,
    },
    selectNode: {
      fontSize: 12,
      top: 0,
      left: 0,
      border: 0,
      width: "100%",
      borderBottom: `0.5px solid ${theme.colors.gray[3]}`,
      marginBottom: 10,
      paddingRight: 20,
    },
    selectPlotType: {
      fontSize: 12,
      position: "absolute",
      top: 0,
      left: 0,
      border: 0,
      width: "100%",
      borderBottom: `0.5px solid ${theme.colors.gray[3]}`,
      marginBottom: 10,
      paddingRight: 10,
    },
    ctrlInputLabel: {
      marginLeft: 10,
    },
    sevenSegmentContainer: {
      display: "flex",
      justifyContent: "center",
      alignItems: "center",
      height: "100%",
      width: "100%",
      paddingBottom: 10,
      minWidth: 480,
      transform: "scale(0.6)",
    },
  };
});

export type ControlComponentProps = {
  ctrlObj: CtlManifestType;
  results: ResultsType;
  updateCtrlValue: (value: string, ctrl: CtlManifestType) => void;
  attachParamsToCtrl: (val: CtrlOptionValue, ctrlObj: CtlManifestType) => void;
  removeCtrl: (
    e: React.MouseEvent<HTMLButtonElement, MouseEvent>,
    ctrl: CtlManifestType
  ) => void;
  setCurrentInput: Dispatch<
    SetStateAction<(CtlManifestType & { index: number }) | undefined>
  >;
  setOpenEditModal: Dispatch<SetStateAction<boolean>>;
};

const ControlComponent = ({
  ctrlObj,
  results,
  updateCtrlValue,
  attachParamsToCtrl,
  removeCtrl,
  setCurrentInput,
  setOpenEditModal,
}: ControlComponentProps) => {
  const { classes } = useControlStyles();
  const theme = useMantineColorScheme().colorScheme;
  const {
    ctrlsManifest,
    setGridLayout,
    isEditMode,
    selectOptions,
    setSelectOptions,
    flowChartObject,
    textInput,
    setTextInput,
    numberInput,
    setNumberInput,
    sliderInput,
    setSliderInput,
    currentInputValue,
    setCurrentInputValue,
    nd,
    setNd,
    setPlotData,
    selectedPlotOption,
    plotData,
    selectedOption,
    setSelectedOption,
    inputOptions,
    outputOptions,
    setOutputOptions,
    setSelectedPlotOption,
    ctrls,
    defaultValue,
    paramOptions,
    styledLayout,
    openFileSelector,
    setInputOptions,
  } = ControlComponentState({
    updateCtrlValue,
    ctrlObj,
  });

  const handleCtrlValueChange = (
    setValue: Dispatch<SetStateAction<string>>,
    value: string
  ) => {
    setValue(value);
    if (!(ctrlObj?.param as CtrlManifestParam)?.nodeId) {
      return;
    }
    updateCtrlValue(value, ctrlObj);

    if ((ctrlObj.param as CtrlManifestParam).functionName === "CONSTANT") {
      attachParamsToCtrl(
        {
          ...(selectedOption?.value as CtrlOptionValue),
          id: `CONSTANT_${value.toString().split(" ").join("")}_constant`,
        },
        { ...ctrlObj, val: value }
      );
    }
  };

  const makeLayoutStatic = () => {
    if (isEditMode) {
      setGridLayout((prev) => {
        prev[prev.findIndex((layout) => layout.i === ctrlObj.id)].static = true;
      });
    }
  };

  useControlComponentEffects({
    nd,
    selectedPlotOption,
    setNd,
    setPlotData,
    ctrlObj,
    ctrls,
    ctrlsManifest,
    currentInputValue,
    defaultValue,
    flowChartObject,
    isEditMode,
    numberInput,
    paramOptions,
    plotData,
    results,
    selectedOption,
    selectOptions,
    inputOptions,
    outputOptions,
    sliderInput,
    styledLayout,
    textInput,
    setCurrentInputValue,
    setGridLayout,
    setNumberInput,
    setSelectedOption,
    setSelectedPlotOption,
    setSelectOptions,
    setOutputOptions,
    setSliderInput,
    setTextInput,
    openFileSelector,
    setInputOptions,
  });
  return (
    <Box
      style={{
        display: "flex",
        flexDirection: "column",
        justifyContent: "center",
        alignItems: "center",
        flex: "1",
      }}
      data-cy={ctrlObj.id}
    >
      {isEditMode && (
        <Box className={classes.header} data-cy="ctrls-select">
          <Select
            className={classes.selectNode}
            isSearchable={true}
            onChange={(val) => {
              if (val)
                attachParamsToCtrl(val.value as CtrlOptionValue, ctrlObj);
            }}
            theme={theme as unknown as ThemeConfig}
            options={selectOptions}
            styles={customDropdownStyles}
            value={selectedOption}
            inputId={`select-input-${ctrlObj.id}`}
          />
          <button
            className={classes.btn}
            onClick={() => {
              setCurrentInput({
                ...ctrlObj,
                index: ctrlsManifest.findIndex(
                  (manifest) => manifest.id === ctrlObj.id
                ),
              });
              setOpenEditModal(true);
            }}
          >
            &#9998;
          </button>

          <button
            onClick={(e) => removeCtrl(e, ctrlObj)}
            id={ctrlObj.id}
            className={clsx(classes.btn, classes.closeBtn)}
          >
            x
          </button>
        </Box>
      )}
      {!isEditMode && (
        <Text
          weight={600}
          sx={{ letterSpacing: 1, fontFamily: "Open Sans" }}
          w="100%"
          ta="left"
        >
          {selectedOption?.label ?? "NODE: "}
        </Text>
      )}

      {/* TODO: Refactor this */}
      {ctrlObj.name === ControlNames.Plot && (
        <PlotControl
          nd={nd}
          setPlotData={setPlotData}
          ctrlObj={ctrlObj}
          isEditMode={isEditMode}
          plotData={plotData}
          selectedPlotOption={selectedPlotOption}
          setSelectedPlotOption={setSelectedPlotOption}
        />
      )}
      {ctrlObj.name === ControlNames.LocalFileLoader && (
        <Box className={classes.inputBodyFile} data-cy="numeric-input">
          <input
            type={"text"}
            placeholder={"Please enter the full file path"}
            className={clsx(classes.ctrlNumericInput, "border-color")}
            onChange={(e) => {
              handleCtrlValueChange(setTextInput, e.target.value);
            }}
            value={currentInputValue || textInput || ""}
          />
          {/* <button
            className="cmd-btn-dark "
            style={{
              ...{color: theme === "dark" ? "#fff" : "#000"}
            }}
            onClick={openFileSelector}
            >
            Select File
          </button> */}
        </Box>
      )}
      {ctrlObj.name === ControlNames.SevenSegmentDisplay && (
        <SevenSegmentComponent ctrlObj={ctrlObj} plotData={plotData} nd={nd} />
      )}

      {ctrlObj.name === ControlNames.TextInput && (
        <Box className={classes.inputBody} data-cy="numeric-input">
          <input
            type="text"
            placeholder="Write your text.."
            className={clsx(classes.ctrlNumericInput, "border-color")}
            onChange={(e) => {
              handleCtrlValueChange(setTextInput, e.target.value);
            }}
            value={currentInputValue || textInput || ""}
          />
        </Box>
      )}

      {ctrlObj.name === ControlNames.NumericInput && (
        <Box className={classes.inputBody} data-cy="numeric-input">
          <input
            type="number"
            placeholder="Enter a number"
            className={clsx(classes.ctrlNumericInput, "border-color")}
            onChange={(e) => {
              handleCtrlValueChange(setNumberInput, e.target.value);
            }}
            value={currentInputValue || numberInput || 0}
          />
        </Box>
      )}
      {ctrlObj.name === ControlNames.ArrayNumericInput && (
        <Box className={classes.inputBody}>
          <input
            type="text"
            placeholder="Enter numbers in CSV format"
            className={clsx(classes.ctrlNumericInput, "border-color")}
            onChange={(e) => {
              handleCtrlValueChange(setTextInput, e.target.value);
            }}
            value={currentInputValue || textInput || ""}
          />
        </Box>
      )}

      {ctrlObj.name === ControlNames.StaticNumericInput && (
        <Box className={classes.inputBody}>
          <input
            type="number"
            placeholder="Enter a number"
            className={classes.ctrlNumericInput}
            onChange={(e) => {
              handleCtrlValueChange(setNumberInput, e.target.value);
            }}
            disabled
            value={currentInputValue || numberInput || 0}
          />
        </Box>
      )}

      {ctrlObj.name === ControlNames.Knob && (
        <Box className={classes.inputBody}>
          <KnobCtrl
            selectedOption={selectedOption}
            ctrlObj={ctrlObj}
            isEditMode={isEditMode}
            makeLayoutStatic={makeLayoutStatic}
            setGridLayout={setGridLayout}
            updateCtrlValue={updateCtrlValue}
            currentInputValue={+currentInputValue}
          />
        </Box>
      )}

      {ctrlObj.name === ControlNames.Slider && (
        <Box className={classes.inputBody}>
          <Box
            onMouseEnter={makeLayoutStatic}
            onMouseLeave={() => {
              if (isEditMode) {
                setGridLayout((prev) => {
                  prev[
                    prev.findIndex((layout) => layout.i === ctrlObj.id)
                  ].static = false;
                });
              }
            }}
            style={{ width: "100%" }}
          >
            <Slider
              className="custom-slider"
              onChange={(val) => {
                handleCtrlValueChange(setSliderInput, val.toString());
              }}
              value={+currentInputValue || +sliderInput || 0}
            />
            <label>{currentInputValue ? currentInputValue : sliderInput}</label>
          </Box>
        </Box>
      )}
      {ctrlObj.name === ControlNames.NodeReference && (
        <NodeReference ctrlObj={ctrlObj} updateCtrlValue={updateCtrlValue} />
      )}
      {ctrlObj.name === ControlNames.Dropdown && (
        <Box className={classes.inputBody}>
          <Select
            className={classes.selectNode}
            isSearchable={true}
            onChange={(val) => {
              updateCtrlValue(
                (val as { label: string; value: string }).value,
                ctrlObj
              );
            }}
            theme={theme as unknown as ThemeConfig}
            options={paramOptions}
            styles={customDropdownStyles}
            value={
              paramOptions.find(
                (opt) => opt.value === currentInputValue.toString()
              ) || { label: "", value: "" }
            }
          />
        </Box>
      )}

      {ctrlObj.name === ControlNames.CheckboxButtonGroup && (
        <Box className={classes.inputBody}>
          {paramOptions.map((option) => {
            return (
              <Box key={option.value}>
                <input
                  type="checkbox"
                  id={`${ctrlObj.id}_${option.value}`}
                  name={`${ctrlObj.id}_${option.value}`}
                  value={option.value}
                  checked={currentInputValue.toString() === option.value}
                  onChange={() => {
                    updateCtrlValue(option.value, ctrlObj);
                  }}
                />
                <label htmlFor={`${ctrlObj.id}_${option.value}`}>
                  {" "}
                  {option.label}{" "}
                </label>
              </Box>
            );
          })}
        </Box>
      )}

      {ctrlObj.name === ControlNames.RadioButtonGroup && (
        <Box className={classes.inputBody}>
          {paramOptions.map((option) => {
            return (
              <Box style={{ width: "max-content" }} key={option.value}>
                <input
                  type="radio"
                  id={`${ctrlObj.id}_${option.value}`}
                  name={`${ctrlObj.id}_${option.value}`}
                  value={option.value}
                  checked={currentInputValue.toString() === option.value}
                  onChange={() => {
                    updateCtrlValue(option.value, ctrlObj);
                  }}
                />
                <label htmlFor={`${ctrlObj.id}_${option.value}`}>
                  {" "}
                  {option.label}{" "}
                </label>
              </Box>
            );
          })}
        </Box>
      )}

      <details className={classes.meta}>
        {`Name: ${ctrlObj.name}`}
        <br></br>
        {`ID: ${ctrlObj.id}`}
      </details>
    </Box>
  );
};
export default memo(ControlComponent);
