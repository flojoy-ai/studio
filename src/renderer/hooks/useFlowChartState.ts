import { ElementsData } from "@/renderer/types";
import { TextData } from "@/renderer/types/node";
import { atom, useAtom } from "jotai";
import { atomWithImmer } from "jotai-immer";
import { ReactFlowJsonObject, Node } from "reactflow";

export interface EnvVarCredentialType {
  key: string;
  value: string;
}

export type Project = {
  name?: string;
  rfInstance?: ReactFlowJsonObject<ElementsData>;
  textNodes?: Node<TextData>[];
};

export const projectAtom = atomWithImmer<Project>({});
export const projectPathAtom = atom<string | undefined>(undefined);
export const showWelcomeScreenAtom = atom<boolean>(true);

export const failedNodeAtom = atom<Record<string, string>>({});
export const runningNodeAtom = atom<string>("");
export const nodeStatusAtom = atom((get) => ({
  runningNode: get(runningNodeAtom),
  failedNodes: get(failedNodeAtom),
}));

const credentialsAtom = atomWithImmer<EnvVarCredentialType[]>([]);
const isSidebarOpenAtom = atom<boolean>(false);
const nodeParamChangedAtom = atom<boolean>(false);
export const centerPositionAtom = atom<{ x: number; y: number }>({
  x: 0,
  y: 0,
});

const currentPythonEnvAtom = atom<string | undefined>(undefined);

export function useFlowChartState() {
  const [runningNode, setRunningNode] = useAtom(runningNodeAtom);
  const [failedNodes, setFailedNodes] = useAtom(failedNodeAtom);
  const [credentials, setCredentials] = useAtom(credentialsAtom);
  const [isSidebarOpen, setIsSidebarOpen] = useAtom(isSidebarOpenAtom);
  const [nodeParamChanged, setNodeParamChanged] = useAtom(nodeParamChangedAtom);
  const [currentPythonEnv, setCurrentPythonEnv] = useAtom(currentPythonEnvAtom);

  return {
    runningNode,
    setRunningNode,
    failedNodes,
    setFailedNodes,
    credentials,
    setCredentials,
    nodeParamChanged,
    setNodeParamChanged,
    isSidebarOpen,
    setIsSidebarOpen,
    currentPythonEnv,
    setCurrentPythonEnv,
  };
}
