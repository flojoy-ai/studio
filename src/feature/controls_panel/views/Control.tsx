import { CtlManifestType } from "@src/hooks/useFlowChartState";
import "react-grid-layout/css/styles.css";
import "../../../App.css";
import "../style/Controls.css";
import { ControlProps } from "../types/ControlProps";
import ControlComponent from "./ControlComponent";

export default function Control({
  controlProps,
  ctrl,
}: {
  controlProps: ControlProps;
  ctrl: CtlManifestType;
  ctrlIndex: number;
}) {
  const {
    isEditMode,
    theme,
    results,
    updateCtrlValue,
    attachParamsToCtrl,
    removeCtrl,
    setCurrentInput,
    setOpenEditModal,
  } = controlProps;

  return (
    <div
      className={isEditMode ? "ctrl-input" : ""}
      data-cy="ctrl-grid-item"
      style={{
        width: "100%",
        height: "100%",
        display: "flex",
        flexDirection: "column",
        borderRadius: "16px",
        backgroundColor: theme === "dark" ? "#14131361" : "#58454517",
      }}
    >
      {isEditMode ? (
        <ControlComponent
          ctrlObj={ctrl}
          theme={theme}
          results={results}
          updateCtrlValue={updateCtrlValue}
          attachParamsToCtrl={attachParamsToCtrl}
          removeCtrl={removeCtrl}
          setCurrentInput={setCurrentInput}
          setOpenEditModal={setOpenEditModal}
        />
      ) : ctrl.hidden ? null : (
        <ControlComponent
          ctrlObj={ctrl}
          theme={theme}
          results={results}
          updateCtrlValue={updateCtrlValue}
          attachParamsToCtrl={attachParamsToCtrl}
          removeCtrl={removeCtrl}
          setCurrentInput={setCurrentInput}
          setOpenEditModal={setOpenEditModal}
        />
      )}
    </div>
  );
}
