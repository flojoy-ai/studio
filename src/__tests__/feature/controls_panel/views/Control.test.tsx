import { getByTestId, render } from "@testing-library/react";
import Control from "@src/feature/controls_panel/views/Control";
import { ControlProps } from "@src/feature/controls_panel/types/ControlProps";
import { CtlManifestType } from "@src/hooks/useFlowChartState";
import ControlComponent from "@src/feature/controls_panel/views/control-component/ControlComponent";

jest.mock("@src/hooks/useFlowChartState");

const controlProps: ControlProps = {
  isEditMode: true,
  theme: "dark",
  results: {},
  updateCtrlValue: jest.fn(),
  attachParamsToCtrl: jest.fn(),
  removeCtrl: jest.fn(),
  setCurrentInput: jest.fn(),
  setOpenEditModal: jest.fn(),
};
const ctrl: CtlManifestType = {
  id: "test-ctrl",
  minHeight: 110,
  minWidth: 150,
  name: "Test ctrl",
  type: "input",
  hidden: false,
  layout: {
    h: 12,
    w: 12,
    i: "test-id",
    x: 10,
    y: 10,
  },
};

// mock `ControlComponent`
jest.mock(
  "@src/feature/controls_panel/views/control-component/ControlComponent",
  () => ({
    __esModule: true,
    default: jest.fn((props) => {
      return (
        <div data-testid="control-component" data-props={JSON.stringify(props)}>
          Control Component
        </div>
      );
    }),
  })
);
describe("Control.tsx", () => {
  it("render Control component correctly.", () => {
    const { container } = render(
      <Control controlProps={controlProps} ctrl={ctrl} />
    );
    expect(getByTestId(container, "ctrl-grid-item")).toHaveClass("ctrl-input");
    expect(ControlComponent).toBeCalled();
    expect(container).toMatchSnapshot();
  });
});
