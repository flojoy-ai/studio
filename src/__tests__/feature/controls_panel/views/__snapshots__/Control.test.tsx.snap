// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Control.tsx render Control component correctly. 1`] = `
<div>
  <div
    class="ctrl-input"
    data-cy="ctrl-grid-item"
    data-testid="ctrl-grid-item"
    style="width: 100%; height: 100%; display: flex; flex-direction: column; border-radius: 16px; background-color: rgba(20, 19, 19, 0.38);"
  >
    <div
      data-props="{"ctrlObj":{"id":"test-ctrl","minHeight":110,"minWidth":150,"name":"Test ctrl","type":"input","hidden":false,"layout":{"h":12,"w":12,"i":"test-id","x":10,"y":10}},"theme":"dark","results":{}}"
      data-testid="control-component"
    >
      Control Component
    </div>
  </div>
</div>
`;
